airflow/utils/serialization.py

```diff
@@ -0,0 +1,291 @@
+
+
+def encode(cls: str, version: int, data: T) -> dict[str, str | int | T]:
+    """Encodes objects so it can be understood by the deserializer"""
+    return {CLASSNAME: cls, VERSION: version, DATA: data}
+
+
+def decode(d: dict[str, str | int | T]) -> tuple:
+    return d[CLASSNAME], d[VERSION], d.get(DATA, None)
+
+
+def serialize(o: object, depth: int = 0) -> U | None:
+    """
+    Recursively serializes objects into a primitive. Primitives (int, float, int, bool)
+    are returned as is. Tuples and dicts are iterated over, where it is assumed that keys
+    for dicts can be represented as str. Values that are not primitives are serialized if
+    a serializer is found for them. The order in which serializers are used
+    is 1) a serialize function provided by the object 2) a registered serializer in
+    the namespace of airflow.utils.serializers and 3) an attr or dataclass annotations.
+    If a serializer cannot be found a TypeError is raised.
+
+    :param o: object to serialize
+    :param depth: private
+    :return: a primitive
+    """
+    if depth == MAX_RECURSION_DEPTH:
+        raise RecursionError("maximum recursion depth reached for serialization")
+
+    # None remains None
+    if o is None:
+        return o
+
+    # primitive types are returned as is
+    if isinstance(o, _primitives):
+        if isinstance(o, enum.IntEnum):
+            return o.value
+
+        return o
+
+    # tuples and plain dicts are iterated over recursively
+    if isinstance(o, _iterables):
+        return [serialize(d, depth + 1) for d in o]
+
+    if isinstance(o, dict):
+        if CLASSNAME in o or SCHEMA_ID in o:
+            raise AttributeError(f"reserved key {CLASSNAME} or {SCHEMA_ID} found in dict to serialize")
+
+        return {str(k): serialize(v, depth + 1) for k, v in o.items()}
+
+    cls = type(o)
+    qn = qualname(o)
+
+    # custom serializers
+    dct = {
+        CLASSNAME: qn,
+        VERSION: getattr(cls, "__version__", DEFAULT_VERSION),
+    }
+
+    if hasattr(o, "serialize"):
+        dct[DATA] = getattr(o, "serialize")()
+        return dct
+
+    # if there is a builtin serializer available use that
+    if qn in _serializers:
+        data, classname, version, is_serialized = _serializers[qn].serialize(o)
+        if is_serialized:
+            return encode(classname, version, data)
+
+    # dataclasses
+    if dataclasses.is_dataclass(cls):
+        data = dataclasses.asdict(o)
+        dct[DATA] = serialize(data, depth + 1)
+        return dct
+
+    # attr annotated classes
+    if attr.has(cls):
+        # Only include attributes which we can pass back to the classes constructor
+        data = attr.asdict(o, recurse=True, filter=lambda a, v: a.init)  # type: ignore[arg-type]
+        dct[DATA] = serialize(data, depth + 1)
+        return dct
+
+    raise TypeError(f"cannot serialize object of type {cls}")
+
+
+def deserialize(o: T | None, full=True, type_hint: Any = None) -> object:
```

Comment:
Since this is only really used by SerializedDAG, I think it’s more productive to simply have `deserialize(...) -> Any` and add a separate `deserialize_dag(...) -> SerializedDAG` that simply calls `deserialize` internally and provide a better return type hint.